pkg/bucket/bandwidth/measurements/bucket_measurement.go

```diff
@@ -0,0 +1,108 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package measurements
+
+import (
+	"math"
+	"sync"
+	"time"
+)
+
+const (
+	weight = 0.5
+)
+
+// BucketMeasurement captures the bandwidth details for one bucket
+type BucketMeasurement struct {
+	lock                 sync.Mutex
+	cumulativeBytes      uint64        // Sum of all bytes transferred before the latest update
+	cumulativeTime       time.Duration // Total time taken for bytes
+	bytesSinceLastWindow uint64        // Total bytes since last window was processed
+	windowStart          time.Time     // Start time for window
+	expMovingAvg         float64       // Previously calculate sliding window
+}
+
+// NewBucketMeasurement creates a new instance of the measurement with the initial start time.
+func NewBucketMeasurement(initTime time.Time) *BucketMeasurement {
+	return &BucketMeasurement{
+		windowStart: initTime,
+	}
+}
+
+// IncrementBytes add bytes reported for a bucket.
+func (m *BucketMeasurement) IncrementBytes(bytes uint64) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	m.cumulativeBytes += bytes
+	m.bytesSinceLastWindow += bytes
+}
+
+// CalculateWindow processes the measurements between fixed time slots.
+func (m *BucketMeasurement) CalculateWindow(windowEnd time.Time) {
+	// Calculate aggregate avg bandwidth and exp window avg
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	defer func() {
+		m.bytesSinceLastWindow = 0
+		m.windowStart = windowEnd
+	}()
+
+	if windowEnd.Before(m.windowStart) {
+		return
+	}
+
+	duration := windowEnd.Sub(m.windowStart)
+
+	// Do not update cumulative time if no traffic
+	if m.bytesSinceLastWindow > 0 {
+		m.cumulativeTime += duration
+	}
+
+	if m.expMovingAvg == 0 {
+		// Should address initial calculation and should be fine for resuming from 0
+		m.expMovingAvg = float64(m.bytesSinceLastWindow) / duration.Seconds()
+		return
+	}
+
+	increment := float64(m.bytesSinceLastWindow) / duration.Seconds()
+	m.expMovingAvg = (1-weight)*increment + weight*m.expMovingAvg
+}
+
+// GetAvgMibps returns the bandwidth in MiBPS
+func (m *BucketMeasurement) GetAvgMibps() float64 {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	seconds := float64(m.cumulativeTime) / float64(time.Second)
+	return float64(m.cumulativeBytes) / (1024 * 1024) / seconds
+}
+
+// GetTotalBytes returns the total bytes observed for a bucket
+func (m *BucketMeasurement) GetTotalBytes() uint64 {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	return m.cumulativeBytes
+}
+
+// GetExpMovingAvgMibps returns the exponential moving average for the bucket in MiBPS
+func (m *BucketMeasurement) GetExpMovingAvgMibps() float64 {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.expMovingAvg == 0 || math.IsNaN(m.expMovingAvg) {
```

Comment:
Must not happen. This will be unrecoverable.

Reply:
Agreed.